File: scala/src/main/scala/tddmicroexercises/braking/BrakingCalculation.scala
package tddmicroexercises.braking

import java.lang.Math


object BrakingCalculation {

    def calculateKineticEnergy(velocity: Double, mass: Double): Double = {
      return mass * velocity * velocity / 2
    }

    def calculateVelocity(energy: Double, mass: Double): Double = {
      return Math.sqrt(2 * energy / mass)
    }

    def calculateStoppingDistance(initialVelocity: Double, mass: Double, brake: Brake): Double = {
      var velocity = initialVelocity
      var distance: Double = 0
      while (velocity > 0) {
        var updatedVelocity: Double = BrakingCalculation.updatedVelocity(velocity, mass, brake)
        if (updatedVelocity < 0) {
          updatedVelocity = 0
        }
        distance += (velocity + updatedVelocity) / 2 * .1
        velocity = updatedVelocity
      }
      return distance
    }

    def updatedVelocity(velocity: Double, mass: Double, brake: Brake): Double = {
      val energyDissipated: Double = brake.powerDissipation(velocity)
      val currentKineticEnergy: Double = calculateKineticEnergy(velocity, mass) - energyDissipated
      if (currentKineticEnergy < 0) {
        return 0
      }
      else {
        return calculateVelocity(currentKineticEnergy, mass)
      }
    }


 }

File: Java/Braking/src/main/java/tddmicroexercises/braking/BrakingCalculation.java
package tddmicroexercises.braking;

public class BrakingCalculation {

    public static double calculateKineticEnergy(double velocity, double mass) {
        return mass*velocity*velocity/2;
    }

    public static double calculateVelocity(double energy, double mass) {
        return Math.sqrt(2*energy/mass);
    }

    public static double calculateStoppingDistance(double velocity, double mass, Brake brake) {
        double distance = 0;
        while (velocity > 0) {

            double updatedVelocity = updatedVelocity(velocity, mass, brake);
            if (updatedVelocity < 0) {
                updatedVelocity = 0;
            }
            distance += (velocity + updatedVelocity)/2*.1;  // one tenth of a second
            velocity = updatedVelocity;
        }
        return distance;
    }


    public static double updatedVelocity(double velocity, double mass, Brake brake) {
        double energyDissipated = brake.powerDissipation(velocity);

        double currentKineticEnergy =  calculateKineticEnergy(velocity, mass) - energyDissipated;
        if (currentKineticEnergy < 0) {
            return 0;
        } else {
            return calculateVelocity(currentKineticEnergy, mass);
        }
    }
}
